src/main/java/org/broadinstitute/hellbender/utils/hmm/segmentation/HiddenMarkovModelPostProcessor.java

```diff
@@ -0,0 +1,1145 @@
+        } else {
+            final Set<STATE> onlyCallState = Collections.singleton(call);
+            final Set<STATE> allStatesExceptCall = new HashSet<>(fbResult.model().hiddenStates());
+            allStatesExceptCall.remove(call);
+            return fbResult.logConstrainedProbability(firstTargetIndex - 1, Arrays.asList(allStatesExceptCall, onlyCallState));
+        }
+    }
+
+    /**
+     * Calculates the probability of a hidden state switch at the end of the segment
+     * from the call state to a different state.
+     *
+     * @param lastTargetIndex segment end target index
+     * @param call the called state
+     * @param fbResult forward-backward result
+     * @param <STATE> type of hidden state
+     * @param <TARGET> type of target
+     * @return a double
+     */
+    public static <STATE, TARGET extends Locatable> double logEndProbability(
+            final int lastTargetIndex,
+            final STATE call,
+            final ForwardBackwardAlgorithm.Result<?, TARGET, STATE> fbResult) {
+        if (lastTargetIndex + 1 == fbResult.positions().size()) {
+            return fbResult.logProbability(lastTargetIndex, call);
+        } else {
+            final Set<STATE> onlyCallState = Collections.singleton(call);
+            final Set<STATE> allStatesExceptCall = new HashSet<>(fbResult.model().hiddenStates());
+            return fbResult.logConstrainedProbability(lastTargetIndex, Arrays.asList(onlyCallState, allStatesExceptCall));
+        }
+    }
+
+    /**
+     * Calculates the posterior expectation of the mean of hidden state values on a segment
+     *
+     * @param firstTargetIndex first target index of the segment
+     * @param segmentLength length of the segment
+     * @param fbResult result of forward-backward algorithm
+     * @param <STATE> type of hidden state; must implement {@link ScalarProvider}
+     * @param <TARGET> type of target; must implement {@link Locatable}
+     * @throws IllegalStateException if segment length is negative
+     * @return segment mean
+     */
+    public static <STATE extends ScalarProvider, TARGET extends Locatable> double calculateSegmentMean(
+            final int firstTargetIndex,
+            final int segmentLength,
+            final ForwardBackwardAlgorithm.Result<?, TARGET, STATE> fbResult) {
+        if (segmentLength < 0) {
+            throw new IllegalArgumentException("Segment length must be positive");
+        }
+        /* trivial case when the segment has length 0 */
+        if (segmentLength == 0) {
+            return Double.NaN;
+        } else {
+            final List<STATE> hiddenStates = fbResult.model().hiddenStates();
+            final double[] hiddenStateValues = hiddenStates.stream()
+                    .mapToDouble(STATE::getScalar)
+                    .toArray();
+            return IntStream.range(firstTargetIndex, firstTargetIndex + segmentLength)
+                    .mapToDouble(targetIndex -> IntStream.range(0, hiddenStates.size())
```

Comment:
Instead of two `mapToDouble` and two sums, what about a single `flatMapToDouble` and a single `sum`?